lp2p/peer.go

```diff
@@ -0,0 +1,386 @@
+		return false
+	}
+
+	return true
+}
+
+func (p *Peer) TrySend(e p2p.Envelope) bool {
+	// todo same as SEND, but if current queue is full (its cap=1), immediately return FALSE
+	if err := p.send(e); err != nil {
+		p.Logger.Error("failed to send message", "channel", e.ChannelID, "method", "TrySend", "err", err)
+		return false
+	}
+
+	return true
+}
+
+func (p *Peer) CloseConn() error {
+	return p.host.Network().ClosePeer(p.addrInfo.ID)
+}
+
+func (p *Peer) send(e p2p.Envelope) (err error) {
+	var (
+		peerID     = p.addrInfo.ID
+		protocolID = ProtocolID(e.ChannelID)
+	)
+
+	payload, err := marshalProto(e.Message)
+	if err != nil {
+		return err
+	}
+
+	var (
+		peerIDStr    = peerID.String()
+		messageType  = protoTypeName(e.Message)
+		payloadLen   = float64(len(payload))
+		metricLabels = []string{
+			"peer_id", peerIDStr,
+			"chID", fmt.Sprintf("%#x", e.ChannelID),
+		}
+
+		// note metric's name is misleading, it's a counter, not sum(bytes_pending)
+		pendingMessagesCounter = p.metrics.PeerPendingSendBytes.With("peer_id", peerIDStr)
+	)
+
+	pendingMessagesCounter.Add(1)
+
+	ctx, cancel := context.WithTimeout(context.Background(), TimeoutStream)
+	defer cancel()
+
+	start := time.Now()
+
+	defer func() {
+		pendingMessagesCounter.Add(-1)
+
+		if err != nil {
+			return
+		}
+
+		p.metrics.PeerSendBytesTotal.With(metricLabels...).Add(payloadLen)
+		p.metrics.MessageSendBytesTotal.With("message_type", messageType).Add(payloadLen)
+
+		p.Logger.Debug(
+			"Sent envelope",
+			"protocol", protocolID,
+			"peer_id", peerIDStr,
+			"send_dur", time.Since(start).String(),
+		)
+	}()
+
+	// if no streams are available, it will block or return an error
+	s, err := p.host.NewStream(ctx, peerID, protocolID)
+	if err != nil {
+		return fmt.Errorf("failed to open stream %s: %w", protocolID, err)
+	}
+
+	return StreamWriteClose(s, payload)
+}
+
+// These methods are not implemented as they're not used by reactors
+// (only by PEX/p2p-transport which is not used with go-libp2p)
+
+func (*Peer) Status() conn.ConnectionStatus { return conn.ConnectionStatus{} }
+func (*Peer) NodeInfo() p2p.NodeInfo        { return nil }
+func (*Peer) RemoteIP() net.IP              { return nil }
+func (*Peer) RemoteAddr() net.Addr          { return nil }
+func (*Peer) IsOutbound() bool              { return false }
+func (*Peer) IsPersistent() bool            { return false }
+func (*Peer) FlushStop()                    {}
+func (*Peer) SetRemovalFailed()             {}
+func (*Peer) GetRemovalFailed() bool        { return false }
+
+// PeerSet represents lazy-initialized peer set adapter for go-libp2p.
+type PeerSet struct {
+	host *Host
+
+	peers map[peer.ID]*Peer
+	mu    sync.RWMutex
+
+	metrics *p2p.Metrics
+	logger  log.Logger
+}
+
+var _ p2p.IPeerSet = (*PeerSet)(nil)
+
+func NewPeerSet(host *Host, metrics *p2p.Metrics, logger log.Logger) *PeerSet {
+	const initialCapacity = 64
```

Comment:
Is this just matching what's already in comet or is there some reason for starting w/ this size?